provisioning/device/devdoc/transport_state_machine_requirements.md

```diff
@@ -0,0 +1,81 @@
+
+
+### register(registrationId: string, authorization: string | X509, requestBody: any, forceRegistration: boolean, callback: Provisioning.ResponseCallback): void
+Register round-trips one step of the registration process, not returning until the operation is complete
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_010: [** `register` shall `connect` the transport if it is not connected. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_011: [** `register` shall fail if the connection fails. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_012: [** `register` shall call `_doFirstRegistrationRequestForFsm`. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_013: [** If the registration request fails, `register` shall fail. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_014: [** If the registration request returns with status==assigned, it shall call `callback` with null, the response body, and the protocol-specific result. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_015: [** If the registration request returns with status==Assigning, it shall begin polling for operation status requests. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_016: [** If the registration request returns with an unknown status, `register` shall fail with a `SyntaxError` and pass the response body and the protocol-specific result to the `callback`. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_017: [** If the registration request returns a successful, `register` shall emit an 'operationStatus' event passing the body of the response. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_018: [** When the polling interval elapses, `register` shall call `_doOperationStatusQueryForFsm`. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_019: [** If `_doOperationStatusQueryForFsm` fails, `register` shall fail. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_020: [** If `_doOperationStatusQueryForFsm` succeeds with status==Assigned, `register` shall complete and pass the body of the response and the protocol-spefic result to the `callback`. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_021: [** If `_doOperationStatusQueryForFsm` succeeds with status==Assigned, `register` shall begin another polling interval. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_022: [** If `_doOperationStatusQueryForFsm` succeeds with an unknown status, `register` shall fail with a `SyntaxError` and pass the response body and the protocol-specific result to the `callback`. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_023: [** If `_doOperationStatusQueryForFsm` succeeds, `register` shall emit an 'operationStatus' event passing the body of the response. **]**
+
+**SRS_NODE_PROVISIONING_TRANSPORT_STATE_MACHINE_18_024: [** If `register` is called while a different request is in progress, it shall fail with an `InvalidOperationError`. **]**
```

Comment:
>InvalidOperationError [](start = 149, length = 21)

Hopefully the error message will be very clear here because the error name doesn't suggest what's going on.

Reply:
I'm using errors.InvalidOperationError('another operation is in progress')

---
In reply to: [146397948](https://github.com/Azure/azure-iot-sdk-node/pull/115#discussion_r146397948) [](ancestors = 146397948)